File: models/stages.py
from typing import List, Optional, Union
from pydantic import BaseModel, HttpUrl, constr
from datetime import datetime


# 🔹 Common Objects
class LatLong(BaseModel):
    lat: float
    long: float
    address: Optional[str]

class EnvironmentalConditions(BaseModel):
    soil_quality: Optional[str]
    moisture: Optional[float]
    temperature: Optional[float]
    humidity: Optional[float]
    weather_conditions: Optional[str]
    irrigation_method: Optional[str]


class FarmingInputs(BaseModel):
    fertilizers: Optional[str]
    pesticides_used: Optional[str]
    organic_certified: bool = False


class PermitCompliance(BaseModel):
    permit_id: str
    permit_type: str
    issuer: str
    valid_until: Optional[datetime]

class FarmerDetails(BaseModel):
    farmer_id: str
    batch_id: str
    crop_id: str
    start_time : datetime

# a. Collection Event
class CollectionEvent(BaseModel):
    batch_id: str
    actor_id: str 
    crop_id: str   
    location: LatLong
    start_date: datetime
    harvest_date: datetime
    environment: Optional[EnvironmentalConditions]
    inputs: Optional[FarmingInputs]
    permits: Optional[List[PermitCompliance]]


#  b. Transport Event
class TransportEvent(BaseModel):
    transport_id: str
    batch_ids: List[int]
    provenance_fhir_url: HttpUrl
    transporter_id: str
    origin: LatLong
    destination: LatLong
    start_time: datetime
    end_time: datetime
    transport_conditions: Optional[EnvironmentalConditions]
    sealed: bool = True
    notes: Optional[str]


#  c. Processing Event
class ProcessingEvent(BaseModel):
    processing_id: str
    batch_id: str
    processing_company_id: str
    company_location: LatLong
    processes_applied: List[str]
    process_conditions: Optional[EnvironmentalConditions]
    start_time: datetime
    end_time: datetime
    visual_inspection: Optional[List[str]]
    equipment_cleaned: bool = True
    notes: Optional[str]


#  d. Quality Test
class TestResults(BaseModel):
    test_id: str
    test_type: str
    value: float
    units: str
    reference_range: Optional[str]
    passed: bool


class QualityTest(BaseModel):
    test_id: str
    batch_id: str
    lab_id: str
    date_of_test: datetime
    test_results: List[TestResults]
    certification_report_url: Optional[HttpUrl]
    notes: Optional[str]


#  Manufacturing Event
class IngredientsModel(BaseModel):
    ingredient_id: str
    name: str
    quantity: float
    units: str


class ManufacturingEvent(BaseModel):
    manufacturing_id: str
    product_name: str
    batch_ids_used: List[int]
    manufacturer_id: str
    manufacture_date: datetime
    ingredients: List[IngredientsModel]
    GMP_compliance: bool = True
    test_ids: Optional[List[str]]
    final_quantity: float
    notes: Optional[str]


#  Packing Event
class PackingEvent(BaseModel):
    packing_id: str
    packing_fhir_url: HttpUrl
    manufacturing_id: str
    packer_id: str
    date_of_packing: datetime
    qr_code_url: Optional[HttpUrl]
    notes: Optional[str]

class QrCodeData(BaseModel):
    from_id: str
    to_id: str  
    crops: int
    from_role: str
    to_role: str
    start_time: Optional[datetime]
    event: Union[CollectionEvent, TransportEvent, ProcessingEvent, QualityTest, ManufacturingEvent, PackingEvent]

File: models/stages.go
package models

import (
	"time"
)

// Common Objects

type LatLong struct {
	Lat     float64 `json:"lat"`
	Long    float64 `json:"long"`
	Address *string `json:"address,omitempty"`
}

type EnvironmentalConditions struct {
	SoilQuality       *string  `json:"soil_quality,omitempty"`
	Moisture          *float64 `json:"moisture,omitempty"`
	Temperature       *float64 `json:"temperature,omitempty"`
	Humidity          *float64 `json:"humidity,omitempty"`
	WeatherConditions *string  `json:"weather_conditions,omitempty"`
	IrrigationMethod  *string  `json:"irrigation_method,omitempty"`
}

type FarmingInputs struct {
	Fertilizers      *string `json:"fertilizers,omitempty"`
	PesticidesUsed   *string `json:"pesticides_used,omitempty"`
	OrganicCertified bool    `json:"organic_certified"`
}

type PermitCompliance struct {
	PermitID   string     `json:"permit_id"`
	PermitType string     `json:"permit_type"`
	Issuer     string     `json:"issuer"`
	ValidUntil *time.Time `json:"valid_until,omitempty"`
}

type FarmerDetails struct {
	ProfileID string    `json:"profile_id"`
	BatchID   string    `json:"batch_id"`
	CropID    string    `json:"crop_id"`
	StartTime time.Time `json:"start_time"`
}

// Collection Event
type CollectionEvent struct {
	BatchID     string                    `json:"batch_id"`
	ProfileID   string                    `json:"profile_id"`
	CropID      string                    `json:"crop_id"`
	Location    LatLong                   `json:"location"`
	StartDate   time.Time                 `json:"start_date"`
	HarvestDate time.Time                 `json:"harvest_date"`
	Environment *EnvironmentalConditions  `json:"environment,omitempty"`
	Inputs      *FarmingInputs            `json:"inputs,omitempty"`
	Permits     *[]PermitCompliance       `json:"permits,omitempty"`
}

// Transport Event
type TransportEvent struct {
	TransportID         string                    `json:"transport_id"`
	BatchID             string                    `json:"batch_id"`
	ProvenanceFhirURL   string                    `json:"provenance_fhir_url"`
	ProfileID           string                    `json:"profile_id"`
	Origin              LatLong                   `json:"origin"`
	Destination         LatLong                   `json:"destination"`
	StartTime           time.Time                 `json:"start_time"`
	EndTime             time.Time                 `json:"end_time"`
	TransportConditions *EnvironmentalConditions  `json:"transport_conditions,omitempty"`
	Sealed              bool                      `json:"sealed"`
	Notes               *string                   `json:"notes,omitempty"`
}

// Processing Event
type ProcessingEvent struct {
	ProcessingID       string                    `json:"processing_id"`
	BatchID            string                    `json:"batch_id"`
	ProfileID          string                    `json:"profile_id"`
	CompanyLocation    LatLong                   `json:"company_location"`
	ProcessesApplied   []string                  `json:"processes_applied"`
	ProcessConditions  *EnvironmentalConditions  `json:"process_conditions,omitempty"`
	StartTime          time.Time                 `json:"start_time"`
	EndTime            time.Time                 `json:"end_time"`
	VisualInspection   *[]string                 `json:"visual_inspection,omitempty"`
	EquipmentCleaned   bool                      `json:"equipment_cleaned"`
	Notes              *string                   `json:"notes,omitempty"`
}

// Quality Test
type TestResults struct {
	TestID         string  `json:"test_id"`
	TestType       string  `json:"test_type"`
	Value          float64 `json:"value"`
	Units          string  `json:"units"`
	ReferenceRange *string `json:"reference_range,omitempty"`
	Passed         bool    `json:"passed"`
}

type QualityTest struct {
	TestID                  string        `json:"test_id"`
	BatchID                 string        `json:"batch_id"`
	ProfileID               string        `json:"profile_id"`
	DateOfTest              time.Time     `json:"date_of_test"`
	TestResults             []TestResults `json:"test_results"`
	CertificationReportURL  *string       `json:"certification_report_url,omitempty"`
	Notes                   *string       `json:"notes,omitempty"`
}

// Manufacturing Event
type IngredientsModel struct {
	IngredientID string  `json:"ingredient_id"`
	Name         string  `json:"name"`
	Quantity     float64 `json:"quantity"`
	Units        string  `json:"units"`
}

type ManufacturingEvent struct {
	ManufacturingID  string             `json:"manufacturing_id"`
	ProductName      string             `json:"product_name"`
	BatchIDsUsed     string             `json:"batch_ids_used"`
	ManufacturerID   string             `json:"manufacturer_id"`
	ManufactureDate  time.Time          `json:"manufacture_date"`
	Ingredients      []IngredientsModel `json:"ingredients"`
	GMPCompliance    bool               `json:"GMP_compliance"`
	TestIDs          *[]string          `json:"test_ids,omitempty"`
	FinalQuantity    float64            `json:"final_quantity"`
	Notes            *string            `json:"notes,omitempty"`
}

// Packing Event
type PackingEvent struct {
	PackingID     string     `json:"packing_id"`
	PackingFhirURL string    `json:"packing_fhir_url"`
	ManufacturingID string   `json:"manufacturing_id"`
	PackerID      string     `json:"packer_id"`
	DateOfPacking time.Time  `json:"date_of_packing"`
	QrCodeURL     *string    `json:"qr_code_url,omitempty"`
	Notes         *string    `json:"notes,omitempty"`
}

// Event interface for polymorphic behavior
type Event interface {
	GetEventType() string
}

// Implement the interface for each event type
func (c CollectionEvent) GetEventType() string   { return "collection" }
func (t TransportEvent) GetEventType() string    { return "transport" }
func (p ProcessingEvent) GetEventType() string   { return "processing" }
func (q QualityTest) GetEventType() string       { return "quality_test" }
func (m ManufacturingEvent) GetEventType() string { return "manufacturing" }
func (p PackingEvent) GetEventType() string      { return "packing" }

type QrCodeData struct {
	FromID    string      `json:"from_id"`
	ToID      string      `json:"to_id"`
	BatchID   string      `json:"batch_id"`
	FromRole  string      `json:"from_role"`
	ToRole    string      `json:"to_role"`
	StartTime *time.Time  `json:"start_time,omitempty"`
	Event     interface{} `json:"event"` // Can hold any of the event types
}
